### PR TITLE
Spruce up the `spark` package

### PR DESCRIPTION
**Pull request by fosskers**
### TODO 

- [x] Surface-level linting / refactors
- [x] Add missing RDD ops

### Motivation

This is necessary to be able to replace the `batch` section of RFML with MAML.

### CODE REVIEW

File: spark/src/main/scala/eval/directive/RDDOpDirectives.scala [hunk 1]
```diff
@@ -1,118 +1,257 @@
...
+
+  /* --- UNARY EXPRESSIONS --- */
+
+  val classify = Directive { case (Classification(_, classMap), RDDResult(r) :: Nil) =>
+    unary({ _.color(ColorMap(classMap.classifications)) }, r)
+  }
+
+  val sin = Directive { case (Sin(_), RDDResult(r) :: Nil) => unary({ _.localMapDouble(math.sin(_)) }, r) }
+  val cos = Directive { case (Cos(_), RDDResult(r) :: Nil) => unary({ _.localMapDouble(math.cos(_)) }, r) }
+  val tan = Directive { case (Tan(_), RDDResult(r) :: Nil) => unary({ _.localMapDouble(math.tan(_)) }, r) }
```
Comment (fosskers) [inline spark/src/main/scala/eval/directive/RDDOpDirectives.scala]: Do these families of functions have to account for the case where the input isn't an RDD? Do we allow these basic mathy ops on Int/Double?

Comment (moradology) [reply spark/src/main/scala/eval/directive/RDDOpDirectives.scala]: Yeah, definitely - it is less pressing for the `RDD` interpreter I suppose - but (e.g.) the local case supports going to either tile or bool for comparisons

Comment (fosskers) [reply spark/src/main/scala/eval/directive/RDDOpDirectives.scala]: Ok, then I think I should add those cases. Feels like the responsible thing to do. Although these will never account for `TileResult` - mixing RDD and `LazyTile` would be weird.

Comment (fosskers) [reply spark/src/main/scala/eval/directive/RDDOpDirectives.scala]: `classify` I'm going to leave as only handling RDDs. Numbers don't make sense there as input.

Comment (moradology) [reply spark/src/main/scala/eval/directive/RDDOpDirectives.scala]: They might, though - consider landcover classification: I could hand you a raster to classify or else a cell's value for classification under whatever scheme I'm working with

File: shared/src/main/scala/ast/MamlKind.scala [hunk 2]
```diff
@@ -1,6 +1,6 @@
 package com.azavea.maml.ast
 
-trait MamlKind
+sealed trait MamlKind
```
Comment (fosskers) [inline shared/src/main/scala/ast/MamlKind.scala:3]: Sealed now.

File: spark/src/main/scala/eval/directive/RDDOpDirectives.scala [hunk 3]
```diff
@@ -1,118 +1,309 @@
...
+  }
+
+  val greaterThanOrEqualTo = Directive { case (GreaterOrEqual(_), res1 :: res2 :: Nil) =>
+    Valid(reduce({_ >= _}, {_ >=: _}, {_ >= _}, {_ >=: _}, {_ >= _}, res1, res2))
+  }
+
+  // TODO The mask can probably be something other than `Polygon`.
+  val masking = Directive {
+    case (Masking(_), RDDResult(r) :: GeomResult(g: Polygon) :: Nil) => Valid(RDDResult(r.mask(g)))
+    case (Masking(_), GeomResult(g: Polygon) :: RDDResult(r) :: Nil) => Valid(RDDResult(r.mask(g)))
```
Comment (fosskers) [inline spark/src/main/scala/eval/directive/RDDOpDirectives.scala]: `Polygon` probably not correct here.

File: spark/src/main/scala/eval/directive/RDDOpDirectives.scala [hunk 4]
```diff
@@ -1,118 +1,309 @@
...
+    val f: (RDD[(SpatialKey, Tile)], RDD[(SpatialKey, Tile)]) => RDD[(SpatialKey, Tile)] = { (a,b) =>
+      a.combineValues(b) { (t1,t2) => t1.combineDouble(t2)(math.pow) }
+    }
+
+    Valid(reduce({_ localPow _}, { (i,r) => r.localPow(i) }, { _ localPow _ }, { (d,r) => r.localPow(d) }, f, res1, res2))
+  }
+
+  /* --- UNARY EXPRESSIONS --- */
+
+  val classify = Directive { case (Classification(_, classMap), RDDResult(r) :: Nil) =>
```
Comment (fosskers) [inline spark/src/main/scala/eval/directive/RDDOpDirectives.scala:230]: Only makes sense for `RDD`.

File: spark/src/main/scala/eval/directive/RDDOpDirectives.scala [hunk 5]
```diff
@@ -1,118 +1,309 @@
...
+  val atan = Directive { case (Atan(_), res :: Nil) =>
+    mathy({_.localMapDouble(math.atan(_))}, { i => DoubleResult(math.atan(i.toDouble)) }, { d => DoubleResult(math.atan(d)) }, res)
+  }
+
+  val floor = Directive { case (Floor(_), res :: Nil) =>
+    mathy({ _.localFloor }, { IntResult(_) }, { d => IntResult(math.floor(d).toInt) }, res)
+  }
+  val ceil = Directive { case (Ceil(_), res :: Nil) =>
+    mathy({ _.localCeil }, { IntResult(_) }, { d => IntResult(math.ceil(d).toInt) }, res)
+  }
```
Comment (fosskers) [inline spark/src/main/scala/eval/directive/RDDOpDirectives.scala:269]: `floor`, `ceil`, and `round` are made to return `IntResult` here, which follows intuition but not Scala. In Scala, `floor: Double => Double`.

File: spark/src/main/scala/eval/directive/RDDOpDirectives.scala [hunk 6]
```diff
@@ -1,118 +1,309 @@
...
+  val undefined = Directive { case (Undefined(_), res :: Nil) =>
+    mathy({ _.localUndefined }, { i => BoolResult(!isData(i)) }, { d => BoolResult(!isData(d)) }, res)
+  }
+
+  val numNegation = Directive { case (NumericNegation(_), res :: Nil) =>
+    mathy({ _.localNegate }, { i => IntResult(i * -1) }, { d => DoubleResult(d * -1) }, res)
+  }
+  val logNegation = Directive { case (LogicalNegation(_), res :: Nil) =>
+    mathy({ _.localNot }, { i => IntResult(if (i == 0) 1 else 0) }, { d => IntResult(if (d == 0) 1 else 0) }, res)
+  }
```
Comment (fosskers) [inline spark/src/main/scala/eval/directive/RDDOpDirectives.scala:300]: Always `IntResult`.

**Review by fosskers — COMMENTED** [review]
Pointing out some things worth noting.

File: spark/src/main/scala/eval/directive/RDDOpDirectives.scala (hunk 3 above)
Comment (moradology) [reply spark/src/main/scala/eval/directive/RDDOpDirectives.scala]: I returned an invalid if i couldn't `.as[MultiPolygon]` for the analogous TMS directive

**Comment by moradology** [comment]
Some of the organizational work around directives in this PR was really useful in the work I was doing as well. +1 and merging

**Pull request merged**